pkg/scheduler/internal/queue/scheduling_queue_test.go

```diff
@@ -858,105 +858,186 @@ func TestPriorityQueue_Pop(t *testing.T) {
+					t.Errorf("Expected: %v after Pop, but got: %v", medPriorityPodInfo.Pod.Name, p.Pod.Name)
+				}
+				err := q.AddUnschedulableIfNotPresent(logger, q.newQueuedPodInfo(medPriorityPodInfo.Pod, "plugin"), q.SchedulingCycle())
+				if err != nil {
+					t.Fatalf("unexpected error from AddUnschedulableIfNotPresent: %v", err)
+				}
+				if len(q.unschedulablePods.podInfoMap) != 1 {
+					t.Error("Expected unschedulablePods to be 1.")
+				}
+				updatedPod := medPriorityPodInfo.Pod.DeepCopy()
+				updatedPod.Annotations["foo"] = "test"
+				return medPriorityPodInfo.Pod, updatedPod
+			},
+			validateFunc: func(q *PriorityQueue, expectedPod *v1.Pod) {
+				rawPodInfo, err := q.podBackoffQ.Pop()
+				podGotFromBackoffQ := rawPodInfo.(*framework.QueuedPodInfo).Pod
+				if err != nil || podGotFromBackoffQ != expectedPod {
+					t.Errorf("Expected: %v after Pop, but got: %v", expectedPod.Name, podGotFromBackoffQ.Name)
+				}
+			},
+		},
+		{
+			name: "updating a pod which is in unschedulable queue, and it is not backing off, it will be moved to active queue",
+			prepareFunc: func(q *PriorityQueue) (oldPod, newPod *v1.Pod) {
+				podInfo := q.newQueuedPodInfo(medPriorityPodInfo.Pod)
+				// To simulate the pod is failed in scheduling in the real world, Pop() the pod from activeQ before testing AddUnschedulableIfNotPresent.
+				err := q.activeQ.Add(podInfo)
+				if err != nil {
+					t.Fatalf("unexpected error from activeQ.Add: %v", err)
+				}
+				if p, err := q.Pop(); err != nil || p.Pod != medPriorityPodInfo.Pod {
+					t.Errorf("Expected: %v after Pop, but got: %v", medPriorityPodInfo.Pod.Name, p.Pod.Name)
+				}
+				// updating a pod which is in unschedulable queue, and it is not backing off,
+				// we will move it to active queue
+				err = q.AddUnschedulableIfNotPresent(logger, q.newQueuedPodInfo(medPriorityPodInfo.Pod, "plugin"), q.SchedulingCycle())
+				if err != nil {
+					t.Fatalf("unexpected error from AddUnschedulableIfNotPresent: %v", err)
+				}
+				if len(q.unschedulablePods.podInfoMap) != 1 {
+					t.Error("Expected unschedulablePods to be 1.")
+				}
+				updatedPod := medPriorityPodInfo.Pod.DeepCopy()
+				updatedPod.Annotations["foo"] = "test1"
+				// Move clock by podInitialBackoffDuration, so that pods in the unschedulablePods would pass the backing off,
+				// and the pods will be moved into activeQ.
+				c.Step(q.podInitialBackoffDuration)
+				return medPriorityPodInfo.Pod, updatedPod
+			},
+			validateFunc: func(q *PriorityQueue, expectedPod *v1.Pod) {
+				if p, err := q.Pop(); err != nil || p.Pod != expectedPod {
+					t.Errorf("Expected: %v after Pop, but got: %v", expectedPod.Name, p.Pod.Name)
+				}
+			},
+		},
+		{
+			name: "updating a pod which is in flightPods and the pod will not be moved to any queue",
+			prepareFunc: func(q *PriorityQueue) (oldPod, newPod *v1.Pod) {
+				q.isSchedulingQueueHintEnabled = true
+				podInfo := q.newQueuedPodInfo(medPriorityPodInfo.Pod)
+				err := q.activeQ.Add(podInfo)
```

Comment:
```suggestion
				// We need to once add this Pod to activeQ and Pop() it so that this Pod is registered correctly in inFlightPods.
				err := q.activeQ.Add(podInfo)
```